faiss_vector_cache.go

```diff
@@ -0,0 +1,195 @@
+	return &vecIndexCache{
+		cache:   make(map[uint16]*cacheEntry),
+		closeCh: make(chan struct{}),
+	}
+}
+
+func (vc *vecIndexCache) Clear() {
+	vc.m.Lock()
+	close(vc.closeCh)
+	vc.cache = nil
+	vc.m.Unlock()
+}
+
+func (vc *vecIndexCache) loadVectorIndex(fieldID uint16,
+	indexBytes []byte) (vecIndex *faiss.IndexImpl, err error) {
+	cachedIndex, present := vc.isIndexCached(fieldID)
+	if present {
+		vecIndex = cachedIndex
+		vc.addRef(fieldID)
+	} else {
+		// if the cache doesn't have vector index, just construct it out of the
+		// index bytes and update the cache.
+		vecIndex, err = faiss.ReadIndexFromBuffer(indexBytes, faiss.IOFlagReadOnly)
+		vc.update(fieldID, vecIndex)
+	}
+	return vecIndex, err
+}
+
+func (vc *vecIndexCache) isIndexCached(fieldID uint16) (*faiss.IndexImpl, bool) {
+	vc.m.RLock()
+	entry, present := vc.cache[fieldID]
+	vc.m.RUnlock()
+	if entry == nil {
+		return nil, false
+	}
+
+	entry.m.RLock()
+	rv := entry.index
+	entry.m.RUnlock()
+
+	return rv, present && (rv != nil)
+}
+
+func (vc *vecIndexCache) addRef(fieldIDPlus1 uint16) {
+	vc.m.RLock()
+	entry := vc.cache[fieldIDPlus1]
+	vc.m.RUnlock()
+
+	entry.addRef()
+}
+
+func (vc *vecIndexCache) refresh() (rv int) {
+	vc.m.Lock()
+	cache := vc.cache
+
+	// for every field reconcile the average with the current sample values
+	for fieldIDPlus1, entry := range cache {
+		sample := atomic.LoadUint64(&entry.tracker.sample)
+		entry.tracker.add(sample)
+		// the comparison threshold as of now is (1 - a). mathematically it
+		// means that there is only 1 query per second on average as per history.
+		// and in the current second, there were no queries performed against
+		// this index.
+		if entry.tracker.avg <= (1 - entry.tracker.alpha) {
+			atomic.StoreUint64(&entry.tracker.sample, 0)
+			entry.closeIndex()
+			delete(vc.cache, fieldIDPlus1)
+			continue
+		}
+		atomic.StoreUint64(&entry.tracker.sample, 0)
+	}
+
+	rv = len(vc.cache)
+	vc.m.Unlock()
+	return rv
+}
+
+func (vc *vecIndexCache) monitor() {
+	ticker := time.NewTicker(1 * time.Second)
+	for {
+		select {
+		case <-vc.closeCh:
+			return
+		case <-ticker.C:
+			numEntries := vc.refresh()
+			if numEntries == 0 {
+				// no entries to be monitored, exit
+				return
+			}
+		}
+	}
+}
+
+func (vc *vecIndexCache) update(fieldIDPlus1 uint16, index *faiss.IndexImpl) {
+	vc.m.Lock()
+
+	// the first time we've hit the cache, try to spawn a monitoring routine
+	// which will reconcile the moving averages for all the fields being hit
+	if len(vc.cache) == 0 {
+		go vc.monitor()
+	}
+
+	_, ok := vc.cache[fieldIDPlus1]
+	if !ok {
+		//  initializing the alpha with 0.4 essentially means that we are favoring
+		// 	the history a little bit more relative to the current sample value.
+		// 	this makes the average to be kept above the threshold value for a
+		// 	longer time and thereby the index to be resident in the cache
+		// 	for longer time.
+		vc.cache[fieldIDPlus1] = initCacheEntry(index, 0.4)
```

Comment:
Shouldn't we be setting ref to 1 here?

Reply:
in the latest patch the ref has been incremented as part of the loadVectorIndex() API (which is parent to this).